src/appleseed.shaders/src/appleseed/as_disney_material.osl

```diff
@@ -265,9 +265,7 @@ shader as_disney_material
 
     if (in_enable_matte)
     {
-        out_outMatteOpacity = in_matte_opacity *
-            in_matte_opacity_color * holdout();
-
+        out_outMatteOpacity = as_matte(in_matte_opacity_color, in_matte_opacity);
```

Comment:
@usakhelo , can you try this in as_plastic.osl ?
```
    int ray_depth = 0;
    int status = getattribute("path:ray_depth", ray_depth);

    if (!status || ray_depth > in_maximum_ray_depth)
    {
        return;
    }

    if (in_enable_matte)
    {
        out_outMatteOpacity = as_matte(in_matte_opacity_color,
                                       in_matte_opacity);
        out_outColor += out_outMatteOpacity;
    }
    
    int compute_transparency = (max(in_transparency) > 0.0) ? 1 : 0;

    if (compute_transparency)
    {
        out_outTransparency = in_transparency * transparent();
        out_outColor += out_outTransparency;
    }

    if (raytype("shadow") || raytype("transparency"))
    {
        return;
    }

    if (in_transparency == 1.0)
    {
        return;
    }

    string distribution;  
```

Computing the matte **before** the transparency call seems to produce the expected results.

Reply:
![screenshot_2018-12-28_13-57-29](https://user-images.githubusercontent.com/387352/50504461-bd67f100-0aa8-11e9-961e-0caab3ae39d3.jpg)

A checkerboard texture is passed to *transparency*. The mirror reflections show the normal (red) shading of the (plastic) sphere, with the checkerboard transparency.
The shadows also take into consideration the transparency.
Matte is set to blue and opacity to 0.5. The alpha channel will be solid 0.5. I think here one has the option to have the matte behaviour be a *solid matte* or a *matte gain* (similar to Maya's).

Reply:
Hi @luisbarrancos ,
I tried this version, but I still had to `return` statement in the matte section, otherwise it shows normal shader. And if there is return statement then transparency is not computed if I understand correctly.

Reply:
@usakhelo i cannot reproduce that. With the following shader the matte is evaluated first, then the transparency, and next if the ray is of type transparency or shadow or transparency is full, the shader exits. You can see in the screenshot the matte color is set to blue, the shadows are transparency (checkerboard pattern) and the reflection of the sphere on the metal plane shows red plastic with checkerboard transparency.

Reply:
@luisbarrancos @est77 I tried it again and it worked! I don't know what happend last time, I probably confused my old and this new shaders.

Reply:
Also transparent objects in front of matte work fine.

Reply:
I did some changes and force-pushed them.

Reply:
Ah, got it. Yes I see some changes and it works fine. If I understand correctly your shader changes are equal to what Luis proposed. Everything works great! Thanks @est77!